src/test/java/com/burgstaller/okhttp/AuthenticationTest.java

```diff
@@ -0,0 +1,341 @@
+                        .whenScenarioStateIs(STARTED)
+                        .willReturn(WireMock.aResponse()
+                                .withHeader("Date", "Wed, 21 Oct 2022 05:21:23 GMT")
+                                .withHeader("Proxy-Authenticate", "Basic realm=\"Access to staging site\"")
+                                .withStatus(407)
+                                .withStatusMessage("Proxy Authentication Required")
+                        ).willSetStateTo(PROXY_AUTHENTICATION_REQUIRED_STATE)
+                );
+        wireMock
+                .register(WireMock.get("/ping").inScenario(scenario)
+                        .whenScenarioStateIs(PROXY_AUTHENTICATION_REQUIRED_STATE)
+                        .withHeader("Proxy-Authorization", containing("Basic cHJveHl1c2VyMTpwcm94eXBhc3N3b3JkMQ=="))
+                        .willReturn(WireMock.aResponse()
+                                .withHeader("Date", "Wed, 21 Oct 2022 05:21:23 GMT")
+                                .withHeader("WWW-Authenticate", "Basic realm=\"Access to staging site\"")
+                                .withStatus(401)
+                                .withStatusMessage("Unauthorized")
+                        ).willSetStateTo(UNAUTHORIZED_STATE)
+                );
+        wireMock
+                .register(WireMock.get("/ping").inScenario(scenario)
+                        .whenScenarioStateIs(UNAUTHORIZED_STATE)
+                        .withHeader("Proxy-Authorization", containing("Basic cHJveHl1c2VyMTpwcm94eXBhc3N3b3JkMQ=="))
+                        .withBasicAuth(username, password)
+                        .willReturn(WireMock.aResponse()
+                                .withHeader("Content/Type", "text/plain")
+                                .withBody(bodyResponse)
+                                .withStatus(200)
+                                .withStatusMessage("OK")
+                        ).willSetStateTo(ACCESS_GRANTED)
+                );
+        wireMock
+                .register(WireMock.get("/ping").inScenario(scenario)
+                        .whenScenarioStateIs(ACCESS_GRANTED)
+                        .withHeader("Proxy-Authorization", containing("Basic cHJveHl1c2VyMTpwcm94eXBhc3N3b3JkMQ=="))
+                        .withBasicAuth(username, password)
+                        .willReturn(WireMock.aResponse()
+                                .withHeader("Content/Type", "text/plain")
+                                .withBody(bodyResponse)
+                                .withStatus(200)
+                                .withStatusMessage("OK")
+                        ).willSetStateTo(ACCESS_GRANTED)
+                );
+        Response response1 = client.newCall(request).execute();
+        assertThat(response1.code()).isEqualTo(200);
+        Response response2 = client.newCall(request).execute();
+        assertThat(response2.code()).isEqualTo(200);
+
```

Comment:
This is very cool. Is there a way to check wiremock that all prepared requests were executed?

Reply:
finally, the last update checks all request and responses, by inspecting events from the wiremock server.